src/fsharp/FSharp.Core/seq.fs

```diff
@@ -854,6 +854,27 @@ namespace Microsoft.FSharp.Collections
         [<CompiledName("Singleton")>]
         let singleton value = mkSeq (fun () -> IEnumerator.Singleton value)
 
+        [<CompiledName("Transpose")>]
+        let transpose (source: seq<seq<'T>>) =
+            checkNonNull "source" source
+            seq { use ie = source.GetEnumerator()
+                  let ra = ResizeArray<_>()
+                  try
+                      yield seq { while ie.MoveNext() do
+                                    let inner = ie.Current.GetEnumerator()
+                                    if inner.MoveNext() then
```

Comment:
If the first ``inner.MoveNext`` is false then ``inner`` never gets added to ``rm`` and thus never gets disposed

Reply:
I feel something is odd about this whole implementation.  If I do a Seq.toArray on the overall result, without iterating each individual sequence, then none of the ``inner`` enumerators have been created, and ``rm`` will not have been populated.  But we will have already executed the outer ``finally` designed to cleanup the ``rm`` array.  This feels wrong.  I suspect you may just have to create a strict data structure and return simpler iterators.

Reply:
All good points, this implementation was kind of a brain dump, while trying not to eagerly evaluate and sequences.  I'll take a look at it later on and try and fix.

Reply:
So I've been playing around with Seq.transpose, something like the following works, but not for infinite lists

```fsharp
    [<CompiledName("Transpose")>]
    let transpose (source: seq<seq<'T>>) =
        checkNonNull "source" source
        source
        |> collect indexed
        |> groupBy fst
        |> map (snd >> (map snd))
```

If we were to allow one dimension of the source to be infinite, which would it be?
`n x inf` or `inf x n`

Reply:
@PatrickMcDonald The above is very natural, I would go with that, at least on first look

Reply:
I think it should be possible to make it work with both infinite dimensions, however it will require maintaining a buffer of enumerators, but I would be an interesting feature.